thavalon-server/src/lobby/mod.rs

```diff
@@ -0,0 +1,270 @@
+                e
+            );
+            let return_err = match e {
+                DBGameError::UpdateError => Err(LobbyError::DatabaseError),
+                DBGameError::InvalidStateError => Err(LobbyError::InvalidStateError),
+                _ => {
+                    log::error!("An unknown error occurred in game {}.", self.friend_code);
+                    Err(LobbyError::UnknownError)
+                }
+            };
+
+            return LobbyResponse::Standard(return_err);
+        }
+        let (sender, receiver) = self.builder.as_mut().unwrap().add_player(display_name);
+
+        let client = PlayerClient {
+            to_game: sender,
+            from_game: Arc::new(Mutex::new(receiver)),
+        };
+
+        self.players.insert(player_id.clone(), client);
+        log::info!(
+            "Successfully added player {} to game {}.",
+            player_id,
+            self.friend_code
+        );
+        LobbyResponse::Standard(Ok(()))
+    }
+
+    /// Updates a player's connections to and from the game and to and from the
+    /// client.
+    async fn update_player_connections(
+        &mut self,
+        player_id: String,
+        ws: WebSocket,
+    ) -> LobbyResponse {
+        log::info!("Updating connections for player {}.", player_id);
+
+        // Get the client and create the four connection endpoints.
+        let client = self.players.get(&player_id).unwrap();
+        let (mut to_player_client, mut from_player_client) = ws.split();
+        let mut to_game = client.to_game.clone();
+        let from_game = client.from_game.clone();
+
+        // Spawn a new task to listen for incoming commands from the player.
+        let client_id = player_id.clone();
+        tokio::task::spawn(async move {
+            while let Some(msg) = from_player_client.next().await {
+                let msg = match msg {
+                    Ok(message) => message.to_str().unwrap().to_owned(),
+                    Err(e) => {
+                        // It's not entirely clear what an error here means, so
+                        // probably best to just close the entire connection.
+                        log::error!(
+                            "An error occurred while waiting for messages from player {}. {}",
+                            client_id,
+                            e
+                        );
+                        break;
+                    }
+                };
+
+                log::debug!("Received message {} from player {}.", msg, client_id);
+                let action: Action = match serde_json::from_str(&msg) {
+                    Ok(action) => action,
+                    Err(e) => {
+                        log::warn!(
+                            "Could not deserialize message from player {}. Message: {}",
+                            client_id,
+                            msg
+                        );
+                        // TODO: For now, just kill the connection.
```

Comment:
That makes sense to me - if the client is sending invalid JSON we probably don't have a great way to recover